src/tests/unit/fixtures/thrift-server/basic_union.strict_union.solution.ts

```diff
@@ -0,0 +1,92 @@
+export type MyUnion = IMyUnionWithField1 | IMyUnionWithField2;
+export interface IMyUnionWithField1 {
+    field1: number;
+    field2?: void;
+}
+export interface IMyUnionWithField2 {
+    field1?: void;
+    field2: thrift.Int64;
+}
```

Comment:
Is this generated with the new strict union mode? If so it's not quite what I would have expected. I thought the initial feature request was to have *discriminated* unions, but there are no type-level tags here. Also it's surprising to me that both fields exist in both interfaces (maybe this is a compatibility thing?).

I was expecting something along the lines of:

```typescript
interface IMyUnionWithField1 {
    kind: "field1";
    field1: number;
}
interface IMyUnionWithField2 {
    kind: "field2";
    field2: thrift.Int64;
}
```

---

BTW, I'm assuming the Thrift IDL for this looked something like the following:

```thrift
union MyUnion {
  1: string field1
  2: i64 field2
}
```

Also, related question: where is the IDL that this was generated from? I can't find it anywhere, so I just guessed.

Reply:
You might actually want to use a field name that's not a valid Thrift identifier for the discriminator (instead of `kind`) to avoid a collision if the union is `union MyUnion { 1: string kind }` or whatever, but that's a minor detail.

Reply:
This is from merging in an old branch. So still WIP. Anyway something like ```interface IMyUnionWithField1 {
    field1: number;
}``` Will cause a compile error if someone then just tries to reference `union.field2` (hence the optional void fields). This enforces use of `kind` to validate which interface you are dealing with. Which may not be a bad thing and is probably pretty standard when receiving objects from a Thrift API. However, when constructing Thrift objects it requires users to be aware of fields that there is this additional property that is not part of the IDL. What I'll be moving toward is accepting objects of the form as currently defined and returning objects with an `__type` field. I'll probably also export the possible values of the `__type` field as constants. Even still one could discriminate off of which fields are defined and undefined. But I'll get to the more elegant form.

Reply:
I don't think enforcing the use of `__type` is necessarily bad. I just think it's in contradiction with the thrift definition. They are defined as structs with optional fields, so I should be able to access the fields. More to the point there's some generic stuff in codegen and the framework that treats structs/unions/exceptions the same and checks fields. All of that would need to be rewritten or worse, just cast to `any`. So the optional null fields satisfy both cases.

I'm definitely going to need to generate some constructor/static method to make these things, even just internally. I'd wanted to stick to interfaces/types, but looks like I'll probably need to generate an implementing class for interface. May be a better way.

Yeah, back to the first point, Thrift unions aren't the same as algebraic types in Haskell or unions in TS they are structs, so follow the same rules as structs.

Another thing that has to be supported is annotations, which is mostly transparent to users at this point, but required by the framework.

I'm going to be futzing with this a little logger than I hoped. Thanks for the feedback.

Reply:
> [Unions] are defined as structs with optional fields, so I should be able to access the fields.

> Thrift unions aren't the same as algebraic types in Haskell or unions in TS they are structs, so follow the same rules as structs

I'm not sure I understand what you mean. The semantics of a Thrift union are different from a Thrift struct—namely you must have exactly one field set and the rest must be undefined. [The spec](https://thrift.apache.org/docs/idl#union) documents this as "[unions] provide a means to transport exactly one field of a possible set of fields, just like union {} in C++".

Were you talking about the wire protocols? Looking at the protocol specs ([1](https://github.com/apache/thrift/blob/master/doc/specs/thrift-protocol-spec.md), [2](https://github.com/apache/thrift/blob/master/doc/specs/thrift-binary-protocol.md)) it does appear that unions and structs are encoded identically ([the binary protocol spec](https://github.com/apache/thrift/blob/master/doc/specs/thrift-binary-protocol.md) calls this out as "a union is encoded exactly the same as a struct with the additional restriction that at most 1 field may be encoded"). But there are a ton of invariants that Thrift IDLs can express (and which are used in codegen) that are not represented at all in wire protocols, e.g. enums, requiredness/optionalness, field names, etc; I don't grok why unions would be treated differently from those things.

> More to the point there's some generic stuff in codegen and the framework that treats structs/unions/exceptions the same and checks fields. All of that would need to be rewritten or worse, just cast to `any`.

This I get. It sounds like it would be a lot of work to fully represent unions as "proper" sum types and there are already runtime checks for this anyway, so the benefit might not be that great. I generally like trying to lift as many invariants up to the type level as possible, but obviously there's comes a point where it's not worth the effort.

(Thanks for the thoughtful responses here, by the way. This is fun stuff to talk about.)